woltka/workflow.py

```diff
@@ -0,0 +1,590 @@
+#!/usr/bin/env python3
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2020--, Qiyun Zhu.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+"""Main classification workflow.
+
+Notes
+-----
+Only in this script can functions directly interface with the user by screen
+output (via `click`) and file input/output, except for raising errors.
+"""
+
+from os import makedirs
+from os.path import join, basename, isfile, isdir
+import click
+
+from .util import (
+    readzip, path2stem, update_dict, allkeys, read_ids, add_dict, intize,
+    delnone, id2file_map, write_map, write_table)
+from .align import Plain, parse_align_file
+from .classify import (
+    assign_none, assign_free, assign_rank, count, strip_index, demultiplex)
+from .tree import (
+    read_names, read_nodes, read_lineage, read_newick, read_ranktb, read_map,
+    fill_root)
+from .ordinal import Ordinal, read_gene_coords, whether_prefix
+
+
+def workflow(input_path, output_path, outmap_dir=None,
+             input_fmt=None, input_ext=None, sample_ids=None, demux=None,
+             ranks=None, above=False, major=None, ambig=True, subok=True,
+             deidx=False, coords_fp=None, overlap=80, names_fp=None,
+             nodes_fp=None, newick_fp=None, lineage_fp=None, ranktb_fp=None,
+             map_fps=[], map_rank=False, lines=1000000):
+    """Main classification workflow which accepts command-line arguments.
+
+    See Also
+    --------
+    .cli.classify
```

Comment:
You should provide an absolute path here
```suggestion
    woltka.cli.classify
```

Reply:
The reason I omitted `woltka` is to allow the program to run without installation :). Meanwhile I don't see any downside of doing `.`. But maybe there are some which I am not aware of. Please let me know if so!

Reply:
I think doing relative imports, e.g., lines 23-32, is fine!

But the way I think about it is that the imports are the computer, and the docstrings are for the user. Unless you expect people to be working in the `woltka` directory, putting `woltka.cli.classify` gives them the path to where they need to go, opposed to `.cli.classify` which kind of tells them "classify is in the cli.py file, which is in whatever directory I am in now"

Reply:
I also did not catch this the first time, but you should probably add all of the parameter descriptions to this docstring.

It would be really annoying to be working in python and trying to find out info about the function, to learn you either have to read code, or open another window to run the cli with `--help`, e.g.,
```
>>> from woltka.workflow import workflow
>>> help(workflow)
Main classification workflow which accepts command-line arguments.
    See Also
    --------
    .cli.classify
        Command-line arguments and help information.
```

Reply:
Re. 1st question: the symbol `.` points to the directory where the Python script is, instead of the current working directory. Therefore, even if you import a function in `workflow.py` from another directory, this `.cli.classify` thing will still work. But if you insist, I will change! Since this is minor.

Re. 2nd question: I guess most users of Woltka are not developers, but people with average bioinformatics background. So the docstring won't be read by anyone but ourselves. That being said, I had docstring for all functions except for this one: as its parameters are explained in `.cli.classify`. Meanwhile, I noted that throughout `workflow.py`, multiple functions are but subsections of `workflow`, in order to make the main function `workflow` less lengthy. These functions all have docstrings, and those are duplicates of the parameters of `workflow` and `.cli.classify`.

In such instance, I am thinking that, have very detailed docstrings for all functions (even duplicates) is great, but for achieve, one needs to copy & paste docstring multiple times, and manually. This will happen every time one modifies the code, and it increases the chance of being inconsistent. I guess (though I don't know the widely accepted answer if any), writing one piece of code only once is favorable. And I guess this applies to docstring as well. So I used `See Also` often to soft-link docstrings.

Does this sound reasonable? Again, I don't think it's major and I am glad to add docstring if you do recommend!

Reply:
Hmm, maybe I am not understanding some of your plans? In the docstring `.cli.classify` will never be evaluated right? For the docstring it is purely informational?
I know that for numpy they have some docstring parsers that would use the relative import information to automatically render some documentation on their website. Was that a use you were intending here?
Otherwise I would think it is better to be more specific.

> writing one piece of code only once is favorable. And I guess this applies to docstring as well.

I would like to challenge this one a little bit. Having more information on your code is better than having less, I'd say. In Numpy docs, `See Also` is usually used to provide more context docstrings, but usually does not entirely replace them. Seeing as `workflow` is the main workhorse (and highest level python accessible function) of this package, it probably good to have more information immediately available.

And just because a developer is comfortable reading code, doesn't mean they want to go on a treasure hunt for your docstring :)

Reply:
Hi @gwarmstrong Thank you for further thoughts! I would like to take your advice and add detailed docstrings to `workflow`!

And btw, there is currently no plan to run [Sphinx](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html) to auto-generate documentations, nor do I plan to run any doctests, so yes, the docstring in Woltka is purely informational. If in the future we want to sell Woltka as a Python library, we can re-address it.